server/aesdsocket.c

```diff
@@ -0,0 +1,317 @@
+
+    // Variable to store size of storage buffer
+    int storage_buffer_size = INITIAL_BUFFER_SIZE;
+
+    // Variable to store return value from recv
+    int bytes_received;
+
+    // Variable to store the value of the iterator within the for loop to check for "\n" character in the while loop
+    int temp_iterator = 0;
+
+    // Flag that is set when \n is received
+    int packet_complete_flag = 0;
+
+    // Variable to track the size of the storage buffer 1 indicates size is 1*1024; used when computing bytes to be written and memcpy operation on the storage buffer
+    int storage_buffer_size_count = 1;
+
+    // Variable that stores number of bytes to be written to the local file after completion of a data packet
+    int bytes_to_write = 0;
+
+    // Variable that stores bytes written to file and is used to read from the file
+    int file_size = 0;
+
+    // String to hold the IP address of the client; used when printing logs
+    char address_string[INET_ADDRSTRLEN];
+
+    while(1) {
+
+        // Continuously restarting connections in the while(1) loop
+        new_fd = accept(socketfd, (struct sockaddr *)&test_addr, &addr_size);
+
+        if (new_fd == -1) {
+            perror("accept");
+            continue;
+        } 
+
+        // Using sock_addr_in and inet_ntop function to print the IP address of the accepted client connection
+        // Reference: https://stackoverflow.com/questions/2104495/extract-ip-from-connection-that-listen-and-accept-in-socket-programming-in-linux
+        struct sockaddr_in *p = (struct sockaddr_in *)&test_addr;
+        syslog(LOG_DEBUG, "Accepted connection from %s", inet_ntop(AF_INET, &p->sin_addr, address_string, sizeof(address_string)));
+
+        // While loop until bytes received is <= 0
+        while ((bytes_received = recv(new_fd, buffer, sizeof(buffer), 0)) > 0 ) {
+
+            printf("Bytes Received = %d\n", bytes_received);
+
+            // Iterating over bytes received to check for "\n" character
+            for (int i = 0; i < bytes_received; i++) {
+                if (buffer[i] == '\n') {
+                    packet_complete_flag = 1;
+                    temp_iterator = i+1;
+                    break;
+                }
+            }
+
+            // Copying from receive buffer to storage buffer 
+            memcpy(storage_buffer + (storage_buffer_size_count - 1)*INITIAL_BUFFER_SIZE, buffer, bytes_received);
+
+            // Check if data packet is complete and assign bytes to write based on temporary iterator
+            if (packet_complete_flag == 1) {
+                bytes_to_write = (storage_buffer_size_count - 1)*INITIAL_BUFFER_SIZE + temp_iterator;
+                break;
+            } else {
+                // Increasing size of storage buffer by INITIAL_BUFFER_SIZE amount if no "\n" character found in buffer
+                char *temp_ptr  = (char *)realloc(storage_buffer, (storage_buffer_size + INITIAL_BUFFER_SIZE));
```

Comment:
@GautamaGandhi I think if you received lots of small buffers which did not include packet terminator you would eventually run out of heap space with this design, since you are mallocing an additional INITIAL_BUFFER_SIZE bytes for each packet regardless of how much space was used.

Reply:
@dwalkes yes, I think I need to add an additional check based on the current buffer size and total bytes received without a "\n". I shall implement this check in the else case prior to assignment 6.